### PR TITLE
Update outbound security groups.

### PR DESCRIPTION
**Pull request by MancunianSam**
Most of the changes here are specifying the destination security groups for postgres and EFS as well as adding in a rule for outbound on 443.
I've changed this to use aws_security_group_rule rather than having the rules specified in the aws_security_group resource because otherwise you get circular dependencies on things like the EFS security groups.

### CODE REVIEW

File: lambda/checksum.tf [hunk 1]
```diff
@@ -78,15 +78,28 @@ resource "aws_security_group" "allow_efs_lambda_checksum" {
...
+  count             = local.count_checksum
+  from_port         = 443
+  protocol          = "tcp"
+  security_group_id = aws_security_group.allow_efs_lambda_checksum[count.index].id
+  to_port           = 443
+  type              = "egress"
+  cidr_blocks       = ["0.0.0.0/0"]
+}
+
+resource "aws_security_group_rule" "efs_rule" {
```
Comment (suzannehamilton) [inline lambda/checksum.tf:97]: How easy would it be to extract small modules for the HTTPS rule and the EFS rule? They're identical for every Lambda except for the security group, which you could pass in as a parameter.

It might help to keep the Lambda modules short and easy to understand.

Comment (MancunianSam) [reply lambda/checksum.tf:97]: I don't think it's going to make it much clearer. Each module needs it's own `main.tf` and `variables.tf`. I did try it and the rule declaration in the checksum file now looks like

```
module allow_https_lambda_checksum_rule {
  source = "./modules/https_security_group_rule"
  security_group_id = aws_security_group.allow_efs_lambda_checksum
  rule_count = local.count_checksum // need this as the lambdas are created conditionally
}

module "allow_efs_lambda_checksum_role" {
  source = "modules/efs_security_group_rule"
  rule_count = local.count_checksum
  security_group_id = aws_security_group.allow_efs_lambda_checksum
  source_security_group = var.efs_security_group_id //need this because efs_security_group_id is passed into a variable to the lambda module
}
```

We'll still have all of the duplication we've currently got because the calls to the modules will be identical apart from the security group and the count parameter and we're only saving 3 or 4 lines so it's not going to make it much easier to read. You're right that there is a lot of duplication but it's probably best to keep all the lambda `.tf` files self contained.

Comment (suzannehamilton) [reply lambda/checksum.tf:97]: Fair enough. I guess the fact that each resource has a name like `allow_efs_lambda_download_files_rule` means that its purpose is clear enough as it is.

Though if we ever do want to build a small module, it doesn't have to have `main.tf` and `variables.tf`. That's just a convention, and if a module is only 20 lines long with a single variable then it might be more readable to put the variable into the same file as the resources.

Comment (MancunianSam) [reply lambda/checksum.tf:97]: Yeah true. I'll leave it as is for the moment then and keep each lambda self contained. Once MVP is done, I would like to have a look at the terraform structure and see if we can clean it up a bit anyway.

File: lambda/templates/export_authoriser_policy.json.tpl [hunk 2]
```diff
@@ -19,6 +19,16 @@
...
       ],
       "Resource": "${kms_arn}"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:CreateNetworkInterface",
+        "ec2:DescribeNetworkInterfaces",
+        "ec2:DeleteNetworkInterface",
+        "elasticfilesystem:ClientWrite"
```
Comment (suzannehamilton) [inline lambda/templates/export_authoriser_policy.json.tpl:29]: What does the export authoriser do with EFS?

File: lambda/templates/api_update.json.tpl [hunk 3]
```diff
@@ -31,6 +31,16 @@
         "kms:Decrypt"
       ],
       "Resource": "${kms_arn}"
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "ec2:CreateNetworkInterface",
```
Comment (suzannehamilton) [inline lambda/templates/api_update.json.tpl:38]: Are these changes needed because the Lambdas have just been moved into the private subnet?

Comment (MancunianSam) [reply lambda/templates/api_update.json.tpl:38]: Yep. Not sure why they're EC2 permissions but you need them to add the lambda to the VPC.

File: lambda/templates/export_authoriser_policy.json.tpl (hunk 2 above)
Comment (MancunianSam) [reply lambda/templates/export_authoriser_policy.json.tpl:29]: Nothing at all :smile: I'll remove that permission.

**Review by suzannehamilton — APPROVED** [review]
